File: apps/web/src/platform/commands/bots.ts
import { hubFetch } from "../http";
import { activeSession } from "../session";
import type { BotProfile } from "@shared/types";
import type {
  ExternalBotRow,
  ExternalBotInviteResult,
  WebhookInfo,
  WebhookCreatedResult,
} from "@wavvon/ui";
import type { BotCapabilityGrants, GameLaunchCard } from "@wavvon/core";

/** GET /admin/bots/external -- admin management view (bots.md §4): every
 *  external bot row (pending invite, active, removed), not just the
 *  member-facing directory `GET /bots` returns. */
export async function adminListExternalBots(): Promise<ExternalBotRow[]> {
  const res = await hubFetch("/admin/bots/external");
  return res.json() as Promise<ExternalBotRow[]>;
}

/** POST /bots -- the real invite-by-pubkey route (bots.md §2, exercised by
 *  ttt-bot/README.md). Maps the hub's `{ invite_token }` response onto the
 *  richer `ExternalBotInviteResult` shape this panel renders. */
export async function adminAddExternalBot(
  pubkey: string,
  localNote: string | null,
): Promise<ExternalBotInviteResult> {
  const res = await hubFetch("/bots", {
    method: "POST",
    body: JSON.stringify({ pubkey, note: localNote }),
  });
  const body = (await res.json()) as { invite_token: string };
  return { bot_invite_token: body.invite_token, pubkey };
}

/** DELETE /bots/:pubkey -- the real bot-removal route (bots.md §2). */
export async function adminRemoveExternalBot(pubkey: string): Promise<void> {
  await hubFetch(`/bots/${pubkey}`, { method: "DELETE" });
}

/** PUT /admin/bots/:pubkey/channels -- replaces channel scope atomically
 *  (bots.md §14). Empty list resets to hub-wide access. */
export async function adminSetBotChannelScope(
  pubkey: string,
  channelIds: string[],
): Promise<void> {
  await hubFetch(`/admin/bots/${pubkey}/channels`, {
    method: "PUT",
    body: JSON.stringify({ channel_ids: channelIds }),
  });
}

/** GET /admin/bots/:pubkey/channels -- current channel scope (bots.md §14).
 *  Empty list means hub-wide access. */
export async function adminGetBotChannelScope(pubkey: string): Promise<string[]> {
  const res = await hubFetch(`/admin/bots/${pubkey}/channels`);
  const body = (await res.json()) as { channel_ids: string[] };
  return body.channel_ids;
}

/** GET /admin/bots/:pubkey/capabilities (bot-capability-layer.md §1). */
export async function adminGetBotCapabilities(pubkey: string): Promise<BotCapabilityGrants> {
  const res = await hubFetch(`/admin/bots/${pubkey}/capabilities`);
  return res.json() as Promise<BotCapabilityGrants>;
}

/** PUT /admin/bots/:pubkey/capabilities -- replaces the granted set atomically. */
export async function adminSetBotCapabilities(pubkey: string, capabilities: string[]): Promise<void> {
  await hubFetch(`/admin/bots/${pubkey}/capabilities`, {
    method: "PUT",
    body: JSON.stringify({ capabilities }),
  });
}

export async function adminListWebhooks(): Promise<WebhookInfo[]> {
  const res = await hubFetch("/admin/webhooks");
  return res.json() as Promise<WebhookInfo[]>;
}

export async function adminCreateWebhook(
  channelId: string,
  displayName: string,
  avatarUrl: string | null,
): Promise<WebhookCreatedResult> {
  const res = await hubFetch("/admin/webhooks", {
    method: "POST",
    body: JSON.stringify({ channel_id: channelId, display_name: displayName, avatar_url: avatarUrl }),
  });
  return res.json() as Promise<WebhookCreatedResult>;
}

export async function adminRegenerateWebhook(webhookId: string): Promise<WebhookCreatedResult> {
  const res = await hubFetch(`/admin/webhooks/${webhookId}`, {
    method: "PATCH",
    body: JSON.stringify({}),
  });
  return res.json() as Promise<WebhookCreatedResult>;
}

export async function adminDeleteWebhook(webhookId: string): Promise<void> {
  await hubFetch(`/admin/webhooks/${webhookId}`, { method: "DELETE" });
}

export interface BotCommandSummary {
  name: string;
  description: string;
}

export interface BotListEntry {
  pubkey: string;
  name: string;
  avatar_url?: string | null;
  description?: string | null;
  /** Profile-declared game descriptor (bot-capability-layer.md §11): drives
   *  the directory card's Play affordance. Absent = bot never declared one. */
  game?: GameLaunchCard | null;
  commands: BotCommandSummary[];
}

export async function listBots(): Promise<BotListEntry[]> {
  const res = await hubFetch("/bots");
  return res.json() as Promise<BotListEntry[]>;
}

export async function listBotCommands(): Promise<Array<{ command: string; description: string; bot_name: string }>> {
  const bots = await listBots();
  return bots.flatMap((b) =>
    b.commands.map((c) => ({ command: c.name, description: c.description, bot_name: b.name }))
  );
}

/** Directory-card lookup for the hover/click bot card (bots.md §10). Sourced
 *  from the same `GET /bots` directory list -- there is no single-bot
 *  profile route, so this filters client-side. */
export async function getBotProfile(pubkey: string): Promise<BotProfile> {
  const bots = await listBots();
  const bot = bots.find((b) => b.pubkey === pubkey);
  if (!bot) throw new Error("Bot not found");
  return {
    pubkey: bot.pubkey,
    name: bot.name,
    avatar_url: bot.avatar_url ?? null,
    description: bot.description ?? null,
    commands: bot.commands,
    game: bot.game ?? null,
  };
}

export function sendComponentInteraction(
  messageId: string,
  customId: string,
  values: string[],
): void {
  const { ws } = activeSession();
  if (ws) {
    ws.send({ type: "component_interaction", message_id: messageId, custom_id: customId, values });
  }
}

/** Joins a bot's mini-app / game-modal session (bot-mini-apps.md, bot-capability-layer.md §2). */
export function sendBotAppJoin(botId: string, channelId: string): void {
  const { ws } = activeSession();
  if (ws) {
    ws.send({ type: "bot_app_join", bot_id: botId, channel_id: channelId });
  }
}

File: apps/desktop/src-tauri/src/bots.rs
use crate::state::{AppState, WsCommand};
use tauri::State;

/// GET /bots -- the member-facing bot directory (bots.md §2). The hub has no
/// bot-creation route at all: a bot is an external Ed25519 identity an admin
/// invites by pubkey, so listing is the only member-level bot call there is.
#[tauri::command]
pub(crate) async fn list_bots(
    hub_url: String,
    state: State<'_, AppState>,
) -> Result<Vec<BotProfileResult>, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .get(format!("{base}/bots"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

// ---------------------------------------------------------------------------
// Component interactions
// ---------------------------------------------------------------------------

#[tauri::command]
pub(crate) async fn send_component_interaction(
    hub_url: String,
    message_id: String,
    custom_id: String,
    values: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let hubs = state.hubs.lock().unwrap();
    let session = hubs
        .values()
        .find(|s| s.hub_url.trim_end_matches('/') == hub_url.trim_end_matches('/'))
        .ok_or_else(|| format!("No active session for hub: {hub_url}"))?;
    let tx = session.ws_tx.clone();
    drop(hubs);
    let payload = serde_json::json!({
        "type": "component_interaction",
        "message_id": message_id,
        "custom_id": custom_id,
        "values": values,
    });
    tx.send(WsCommand::Raw(payload.to_string()))
        .map_err(|_| "WS closed".to_string())
}

// ---------------------------------------------------------------------------
// Bot profile
// ---------------------------------------------------------------------------

#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct BotCommandDef {
    pub name: String,
    pub description: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct BotProfileResult {
    pub pubkey: String,
    pub name: String,
    pub avatar_url: Option<String>,
    pub description: Option<String>,
    pub commands: Vec<BotCommandDef>,
}

/// The hub exposes no single-bot route -- `/bots/{pubkey}` is DELETE only --
/// so the profile card is filtered out of the directory list, as the web
/// client does.
#[tauri::command]
pub(crate) async fn get_bot_profile(
    hub_url: String,
    pubkey: String,
    state: State<'_, AppState>,
) -> Result<BotProfileResult, String> {
    list_bots(hub_url, state)
        .await?
        .into_iter()
        .find(|b| b.pubkey == pubkey)
        .ok_or_else(|| "Bot not found".to_string())
}

// ---------------------------------------------------------------------------
// External bots
// ---------------------------------------------------------------------------

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub(crate) struct ExternalBotRow {
    pub public_key: String,
    pub display_name: Option<String>,
    pub local_note: Option<String>,
    pub approval_status: String,
    pub last_seen_at: Option<i64>,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct ExternalBotInviteResult {
    pub bot_invite_token: String,
    pub pubkey: String,
}

#[tauri::command]
pub(crate) async fn admin_list_external_bots(
    hub_url: String,
    state: State<'_, AppState>,
) -> Result<Vec<ExternalBotRow>, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .get(format!("{base}/admin/bots/external"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

#[tauri::command]
pub(crate) async fn admin_add_external_bot(
    hub_url: String,
    pubkey: String,
    local_note: Option<String>,
    state: State<'_, AppState>,
) -> Result<ExternalBotInviteResult, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .post(format!("{base}/bots"))
        .bearer_auth(&token)
        .json(&serde_json::json!({ "pubkey": pubkey, "note": local_note }))
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    #[derive(serde::Deserialize)]
    struct InviteBotResponse {
        invite_token: String,
    }
    let body: InviteBotResponse = resp
        .json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))?;
    Ok(ExternalBotInviteResult {
        bot_invite_token: body.invite_token,
        pubkey,
    })
}

#[tauri::command]
pub(crate) async fn admin_remove_external_bot(
    hub_url: String,
    pubkey: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .delete(format!("{base}/bots/{pubkey}"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    Ok(())
}

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub(crate) struct BotChannelScopeResult {
    pub channel_ids: Vec<String>,
}

#[tauri::command]
pub(crate) async fn admin_get_bot_channel_scope(
    hub_url: String,
    pubkey: String,
    state: State<'_, AppState>,
) -> Result<Vec<String>, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .get(format!("{base}/admin/bots/{pubkey}/channels"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    let parsed: BotChannelScopeResult = resp
        .json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))?;
    Ok(parsed.channel_ids)
}

#[tauri::command]
pub(crate) async fn admin_set_bot_channel_scope(
    hub_url: String,
    pubkey: String,
    channel_ids: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .put(format!("{base}/admin/bots/{pubkey}/channels"))
        .bearer_auth(&token)
        .json(&serde_json::json!({ "channel_ids": channel_ids }))
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Webhooks
// ---------------------------------------------------------------------------

#[derive(serde::Serialize, serde::Deserialize, Clone)]
pub(crate) struct WebhookInfo {
    pub id: String,
    pub display_name: String,
    pub channel_id: String,
    pub channel_name: Option<String>,
    pub webhook_url: String,
    pub created_by: String,
    pub created_at: i64,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) struct WebhookCreatedResult {
    pub id: String,
    pub webhook_url: String,
}

#[tauri::command]
pub(crate) async fn admin_list_webhooks(
    hub_url: String,
    state: State<'_, AppState>,
) -> Result<Vec<WebhookInfo>, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .get(format!("{base}/admin/webhooks"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

#[tauri::command]
pub(crate) async fn admin_create_webhook(
    hub_url: String,
    channel_id: String,
    display_name: String,
    avatar_url: Option<String>,
    state: State<'_, AppState>,
) -> Result<WebhookCreatedResult, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .post(format!("{base}/admin/webhooks"))
        .bearer_auth(&token)
        .json(&serde_json::json!({
            "channel_id": channel_id,
            "display_name": display_name,
            "avatar_url": avatar_url,
        }))
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

#[tauri::command]
pub(crate) async fn admin_regenerate_webhook(
    hub_url: String,
    webhook_id: String,
    state: State<'_, AppState>,
) -> Result<WebhookCreatedResult, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .patch(format!("{base}/admin/webhooks/{webhook_id}"))
        .bearer_auth(&token)
        .json(&serde_json::json!({}))
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

#[tauri::command]
pub(crate) async fn admin_delete_webhook(
    hub_url: String,
    webhook_id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .delete(format!("{base}/admin/webhooks/{webhook_id}"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    Ok(())
}

/// GET /admin/bots/:pubkey/capabilities — the requested, granted and effective
/// sets for one bot (bot-capability-layer.md §1).
#[tauri::command]
pub(crate) async fn admin_get_bot_capabilities(
    hub_url: String,
    pubkey: String,
    state: State<'_, AppState>,
) -> Result<serde_json::Value, String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .get(format!("{base}/admin/bots/{pubkey}/capabilities"))
        .bearer_auth(&token)
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    resp.json()
        .await
        .map_err(|e| format!("Invalid response: {e}"))
}

/// PUT /admin/bots/:pubkey/capabilities — replaces the granted set atomically.
#[tauri::command]
pub(crate) async fn admin_set_bot_capabilities(
    hub_url: String,
    pubkey: String,
    capabilities: Vec<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let token = crate::state::session_for_url(&state, &hub_url)?;
    let base = hub_url.trim_end_matches('/');
    let resp = state
        .http_client
        .put(format!("{base}/admin/bots/{pubkey}/capabilities"))
        .bearer_auth(&token)
        .json(&serde_json::json!({ "capabilities": capabilities }))
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;
    if !resp.status().is_success() {
        return Err(resp.text().await.unwrap_or_default());
    }
    Ok(())
}
